### PR TITLE
Add test for portfolio optimizers.

### PR DESCRIPTION
**Pull request by nalzok**
## Description

Add test for portfolio optimizers. The tests double as alternative implementations.

## Motivation and Context

Currently there are no tests for the portfolio optimizers, so I wrote some for you.

In addition, in addition, the implementations with `cvxopt` in my tests appear to reach a lower loss than the current ones with `scipy.optimize` :).

## How Has This Been Tested?
- [ ] Pass the test by running: `pytest qlib/tests/test_all_pipeline.py` under upper directory of `qlib`.
- [x] If you are adding a new feature, test on your own test scripts.

## Screenshots of Test Results (if appropriate):
1. Pipeline test:
2. Your own tests:

## Types of changes

- [x] Fix bugs
- [ ] Add new feature
- [ ] Update documentation

### CODE REVIEW

**Comment by evanzd** [comment]
@nalzok Thanks for your interest with the portfolio module! 😄 

As the portfolio module is still under development and the docs are not ready yet, there may be some misunderstanding with the parameter `delta`. In fact, it represents the turnover limit `|w - w0|_1 <= delta`: https://github.com/microsoft/qlib/blob/main/qlib/portfolio/optimizer.py#L234. Besides, the alpha vector `u` will be normalized to match the volatility in the covariance matrix: https://github.com/microsoft/qlib/blob/main/qlib/portfolio/optimizer.py#L97.

After fixing these issues, the `GMV` optimizer should give the optimal results (I use CVXPY in the following example to demonstrate this). 
However, as the `MVO` optimizer is sensitive to the inputs, `scipy` failed to solve the optimization problem. We will investigate this issue in the future (maybe will use cvxpy/cvxopt if the problem is already convex).

```python
import unittest

import cvxpy as cp
import numpy as np

from qlib.portfolio.optimizer import PortfolioOptimizer


class TestOptimizer(unittest.TestCase):
    def setUp(self) -> None:
        print('In method', self._testMethodName)

        np.random.seed(42)
        self.S = np.cov(np.random.random((50, 1024)))
        u = np.random.random(50)
        # NOTE: Qlib will scale returns to match the volatility of covariance
        # https://github.com/microsoft/qlib/blob/main/qlib/portfolio/optimizer.py#L97
        self.u = u / u.std() * np.mean(np.diag(self.S)) ** 0.5
        self.w0 = np.ones(50) / 50

    def test_gmv(self):
        delta = 0.6
        alpha = 1.0

        optimizer = PortfolioOptimizer('gmv', delta=delta, alpha=alpha)
        w_qlib = optimizer(self.S, w0=self.w0)

        w = cp.Variable(len(self.w0), nonneg=True)
        obj = cp.Minimize(cp.quad_form(w, self.S) + alpha * cp.norm(w) ** 2)
        cons = [
            cp.sum(w) == 1,
            # NOTE: the sum of changed weights <= delta
            # https://github.com/microsoft/qlib/blob/main/qlib/portfolio/optimizer.py#L234
            cp.norm(w - self.w0, 1) <= delta
        ]
        prob = cp.Problem(obj, cons)
        prob.solve(solver=cp.ECOS)
        self.assertEqual(prob.status, 'optimal')
        w_cvxpy = w.value

        for algo, w in [('Initial', self.w0),
                        ('Qlib', w_qlib),
                        ('CVXPY', w_cvxpy)]:
            self.assertTrue(np.all(w <= 1.), msg=algo)
            self.assertTrue(np.all(0. <= w), msg=algo)
            self.assertAlmostEqual(w.sum(), 1., msg=algo)
            self.assertLessEqual(np.sum(np.abs(w - self.w0)), delta, msg=algo)
            loss = alpha * w.T @ w + w.T @ self.S @ w
            print(f'algo={algo}, loss={loss}')

    def test_mvo(self):
        lamb = 5.0
        delta = 0.6
        alpha = 0.1

        optimizer = PortfolioOptimizer('mvo', lamb=lamb, delta=delta, alpha=alpha)
        w_qlib = optimizer(self.S, self.u, self.w0)

        w = cp.Variable(len(self.w0), nonneg=True)
        obj = cp.Minimize(lamb * cp.quad_form(w, self.S) + alpha * cp.norm(w) ** 2 - w.T @ self.u)
        cons = [
            w >= 0,
            cp.sum(w) == 1,
            # NOTE: the sum of changed weights <= delta
            # https://github.com/microsoft/qlib/blob/main/qlib/portfolio/optimizer.py#L234
            cp.norm(w - self.w0, 1) <= delta
        ]
        prob = cp.Problem(obj, cons)
        prob.solve(solver=cp.ECOS)
        self.assertEqual(prob.status, 'optimal')
        w_cvxpy = w.value

        for algo, w in [('Initial', self.w0),
                        ('Qlib', w_qlib),
                        ('CVXPY', w_cvxpy)]:
            self.assertTrue(np.all(w <= 1.), msg=algo)
            self.assertTrue(np.all(0. <= w), msg=algo)
            self.assertAlmostEqual(w.sum(), 1., msg=algo)
            self.assertLessEqual(np.sum(np.abs(w - self.w0)), delta, msg=algo)
            loss = alpha * w.T @ w - w.T @ self.u + lamb * w.T @ self.S @ w
            print(f'algo={algo}, loss={loss}')


if __name__ == "__main__":
    unittest.main()
```

**Comment by evanzd** [comment]
Close this PR as we are working on adding tests/example/docs for the portfolio module.
